miio/push_server.py

```diff
@@ -0,0 +1,401 @@
+
+    def _create_udp_server(self):
+        """Create the UDP socket and protocol."""
+        # Connect to the miio device to get server_ip using a one time use socket
+        get_ip_socket = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)
+        get_ip_socket.bind((self._address, SERVER_PORT))
+        get_ip_socket.connect((self._device_ip, SERVER_PORT))
+        self._server_ip = get_ip_socket.getsockname()[0]
+        get_ip_socket.close()
+        _LOGGER.debug("Miio push server device ip=%s", self._server_ip)
+
+        # Create a fresh socket that will be used for the push server
+        udp_socket = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)
+        udp_socket.bind((self._address, SERVER_PORT))
+
+        loop = asyncio.get_event_loop()
+
+        return loop.create_datagram_endpoint(
+            lambda: self.ServerProtocol(loop, udp_socket, self),
+            sock=udp_socket,
+        )
+
+    def register_miio_device(self, device: Device, callback):
+        """Register a miio device to this push server."""
+        if device.ip is None:
+            _LOGGER.error(
+                "Can not register miio device to push server since it has no ip"
+            )
+            return
+        if device.token is None:
+            _LOGGER.error(
+                "Can not register miio device to push server since it has no token"
+            )
+            return
+
+        event_ids = []
+        if device.ip in self._registered_devices:
+            _LOGGER.error(
+                "A device for ip '%s' was already registed, overwriting previous callback",
+                device.ip,
+            )
+            event_ids = self._registered_devices[device.ip]["event_ids"]
+
+        self._registered_devices[device.ip] = {
+            "callback": callback,
+            "token": bytes.fromhex(device.token),
+            "event_ids": event_ids,
+            "device": device,
+        }
+
+    def unregister_miio_device(self, device: Device):
+        """Unregister a miio device from this push server."""
+        if device.ip is None:
+            return
+
+        if device.ip in self._registered_devices:
+            for event_id in self._registered_devices[device.ip]["event_ids"]:
+                self.unsubscribe_event(device, event_id)
+            self._registered_devices.pop(device.ip)
+            _LOGGER.debug("push server: unregistered miio device with ip %s", device.ip)
+
+    async def start(self):
+        """Start Miio push server."""
+        if self._listen_couroutine is not None:
+            _LOGGER.error("Miio push server already started, not starting another one.")
+            return
+
+        listen_task = self._create_udp_server()
+        _, self._listen_couroutine = await listen_task
+
+    def stop(self):
+        """Stop Miio push server."""
+        if self._listen_couroutine is None:
+            return
+
+        for ip in list(self._registered_devices):
```

Comment:
Actually there is: the self._registered_devices will be altered inside the for loop, this will result in a dict changed size exception. The list() ensures a copy of the keys is made such that the lenght of that coped list does not changed while itterating though the for loop.